src/zip.c

```diff
@@ -1654,6 +1656,73 @@ ssize_t zip_entry_noallocread(struct zip_t *zip, void *buf, size_t bufsize) {
   return (ssize_t)zip->entry.uncomp_size;
 }
 
+ssize_t zip_entry_noallocread_offset(struct zip_t *zip,
+                                    size_t offset, size_t size, void *buf) {
+  mz_zip_archive *pzip = NULL;
+
+  if (!zip) {
+    // zip_t handler is not initialized
+    return (ssize_t)ZIP_ENOINIT;
+  }
+
+  if (offset > (size_t)zip->entry.uncomp_size) {
+    return (ssize_t)ZIP_EINVAL;
+  }
+
+  if ((offset+size) > (size_t)zip->entry.uncomp_size) {
+    size = (ssize_t)zip->entry.uncomp_size - offset;
+  }
+
+  pzip = &(zip->archive);
+  if (pzip->m_zip_mode != MZ_ZIP_MODE_READING ||
+      zip->entry.index < (ssize_t)0) {
+    // the entry is not found or we do not have read access
+    return (ssize_t)ZIP_ENOENT;
+  }
+
+  mz_zip_reader_extract_iter_state* iter =
+      mz_zip_reader_extract_iter_new(pzip, (mz_uint)zip->entry.index, 0);
+  if (!iter) {
+    return (ssize_t)ZIP_ENORITER;
+  }
+
+  mz_uint8  tmpbuf[ZIP_DEFAULT_ITER_BUF_SIZE];
+  size_t    tmpbuf_size = sizeof(tmpbuf);
+  size_t    file_offset = 0;
+  size_t    write_cursor = 0;
+  size_t    to_read = size;
+
+  // iterate until the requested offset is in range
+  while (file_offset < zip->entry.uncomp_size && to_read > 0)
+  {
+    size_t nread = mz_zip_reader_extract_iter_read(iter, tmpbuf, tmpbuf_size);
+
+    if (nread == 0)
+      break;
+
+    if (offset < (file_offset+nread)) {
+      size_t read_cursor = offset - file_offset;
+      MZ_ASSERT(read_cursor < tmpbuf_size);
+      size_t read_size = nread - read_cursor;
+
+      if (to_read < read_size)
+        read_size = to_read;
+      MZ_ASSERT(read_size <= tmpbuf_size);
+
+      memcpy(&((mz_uint8*)buf)[write_cursor], &tmpbuf[read_cursor], read_size);
```

Comment:
Do we really need this _tmpbuf_, and `memcpy` here?
`mz_zip_reader_extract_iter_read` already calls `memcpy`. Maybe we can optimize it a litle bit, if we can get rid of `tmpbuf`.

Reply:
Well, the question is, where should mz_zip_reader_extract_iter_read() copy to?

Perhaps you meant that the caller also provides the tempbuffer?
Something like:
```c
ssize_t zip_entry_noallocread_offset(struct zip_t *zip,
                                    size_t offset, size_t size, void *buf,
                                    size_t tmpbufsize, void *tmpbuf)
```

Reply:
I get what you mean now. 
I think it should be possible.
It will likely still need a memmove, to put any "unaligned" bytes from the middle of the buffer, into the start of the buffer.